lib/ansible/modules/cloud/amazon/s3_bucket.py

```diff
@@ -196,10 +202,32 @@ def hashable_policy(policy, policy_list):
     if len(policy_list) == 1 and isinstance(policy_list[0], tuple):
         policy_list = policy_list[0]
     if isinstance(policy_list, list):
-        policy_list.sort()
+        if PY3_COMPARISON:
+            policy_list.sort(key=cmp_to_key(py3cmp))
+        else:
+            policy_list.sort()
     return policy_list
 
 
+def py3cmp(a, b):
+    ''' Python 2 can sort lists of mixed types. Strings < tuples. Without this function this fails on Python 3.'''
+    try:
+        if a > b:
+            return 1
+        elif a < b:
+            return -1
+        else:
+            return 0
+    except TypeError as e:
+        # check to see if they're tuple-string
+        # always say strings are less than tuples (to maintain compatibility with python2)
+        if 'unorderable types: str() > tuple()' == to_text(e):
+            return -1
+        elif 'unorderable types: tuple() > str()' == to_text(e):
```

Comment:
@abadger Yeah, I'm not really sure how to go about this.  What I'm doing feels ugly. Using isinstance() was the first approach I took. The issue is that a and b are both actually tuples and are at some point being iterated over (and they don't have the same number tuples contained, leading an inner string to be compared with an inner tuple). An example might look like: ('Action', ('s3:PutObjectAcl',)) and the other might look like: ('Action', (('s3:PutObject',), ('s3:PutObjectAcl',))).
In action:
```
>>> li = [('Action', ('s3:PutObjectAcl',)), ('Action', (('s3:PutObject',), ('s3:PutObjectAcl',)))]
>>> def py3cmp(a, b):
...     try:
...         if a > b:
...             return 1
...         elif a < b:
...             return -1
...         else:
...             return 0
...     except TypeError as e:
...         print("Error: %s" % e)
...         if isinstance(a, string_types) and isinstance(b, tuple):
...             return -1
...         elif isinstance(a, tuple) and isinstance(b, string_types):
...             return 1
...         raise
...
>>>
>>> from functools import cmp_to_key
>>> sorted_list = sorted(li, key=cmp_to_key(py3cmp))
Error: unorderable types: tuple() > str()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in py3cmp
TypeError: unorderable types: tuple() > str()
```